### ISSUE TITLE
Role mapping across trusted clusters does not work if `{{internal.kubernetes_groups}}` trait is the only one specified

### ISSUE THREAD

**Issue by webvictim** [gravitational/teleport#3376]
**What happened**: Reported by a customer and verified by me.

When trying to log into a leaf cluster which has `kubernetes_groups` defined from a root cluster which does not, the role mapping is not processed correctly.

`example-main` = root cluster
`example-east` = leaf cluster

Log into root cluster:
```bash
$ tsh login --proxy=example-main.gravitational.co
> Profile URL:  https://example-main.gravitational.co:3080
  Logged in as: gus@gravitational.com
  Cluster:      example-main.gravitational.co
  Roles:        dev*
  Logins:       root
  Valid until:  2020-02-24 23:58:14 -0400 AST [valid for 12h0m0s]
  Extensions:   permit-agent-forwarding, permit-port-forwarding, permit-pty


* RBAC is only available in Teleport Enterprise
  https://gravitational.com/teleport/docs/enterprise
```

List nodes:
```bash
$ tsh ls
Node Name       Address        Labels                                                   
--------------- -------------- -------------------------------------------------------- 
example-main-node-0 10.4.0.24:3022 environment=demo, kernel=4.14.138+                       
                               role=node, uptime=up 1 week, 2 days, 5 hours, 40 minutes 
example-main-node-1 10.4.0.77:3022 environment=demo, kernel=4.14.138+                       
                               role=node, uptime=up 1 week, 2 days, 5 hours, 39 minutes 
example-main-node-2 10.4.0.55:3022 environment=demo, kernel=4.14.138+                       
                               role=node, uptime=up 1 week, 2 days, 5 hours, 39 minutes 
```

Check Kubernetes context - see root context is shown and selected:
```bash
$ kubectl config get-contexts
CURRENT   NAME                            CLUSTER                                                   AUTHINFO                                              NAMESPACE
*         example-main.gravitational.co   example-main.gravitational.co                             example-main.gravitational.co                             
```

Get pods - this request is expected to fail as there are no groups defined on the root cluster:

```bash
$ kubectl get pods
error: the server doesn't have a resource type "pods"
```

Now, log into the leaf cluster via the root cluster:

```bash
$ tsh login --proxy=example-main.gravitational.co example-east.gravitational.co
> Profile URL:  https://example-main.gravitational.co:3080
  Logged in as: gus@gravitational.com
  Cluster:      example-east.gravitational.co
  Roles:        dev*
  Logins:       root
  Valid until:  2020-02-24 23:58:14 -0400 AST [valid for 11h59m0s]
  Extensions:   permit-agent-forwarding, permit-port-forwarding, permit-pty


* RBAC is only available in Teleport Enterprise
  https://gravitational.com/teleport/docs/enterprise
```

List Kubernetes contexts - see that the leaf context is now active:
```bash
$ kubectl config get-contexts
CURRENT   NAME                            CLUSTER                                                   AUTHINFO                                              NAMESPACE
*         example-east.gravitational.co   example-east.gravitational.co                             example-east.gravitational.co                             
          example-main.gravitational.co   example-main.gravitational.co                             example-main.gravitational.co                             
```

Get pods - this should work, as the leaf server should be given the `system:masters` group, but it does not:

```bash
$ kubectl -v=7 get pods
I0224 13:11:02.442817    4303 loader.go:375] Config loaded from file:  /home/gus/.kube/config
I0224 13:11:02.448667    4303 round_trippers.go:420] GET https://example-main.gravitational.co:3026/api/v1/namespaces/default/pods?limit=500
I0224 13:11:02.448681    4303 round_trippers.go:427] Request Headers:
I0224 13:11:02.448687    4303 round_trippers.go:431]     Accept: application/json;as=Table;v=v1beta1;g=meta.k8s.io, application/json
I0224 13:11:02.448692    4303 round_trippers.go:431]     User-Agent: kubectl/v1.16.2 (linux/amd64) kubernetes/c97fe50
I0224 13:11:02.660138    4303 round_trippers.go:446] Response Status: 403 Forbidden in 211 milliseconds
I0224 13:11:02.660344    4303 request.go:1171] body was not decodable (unable to check for Status): Object 'Kind' is missing in '{"message":"[00] access denied"}'
I0224 13:11:02.660532    4303 helpers.go:199] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "unknown (get pods)",
  "reason": "Forbidden",
  "details": {
    "kind": "pods",
    "causes": [
      {
        "reason": "UnexpectedServerResponse",
        "message": "unknown"
      }
    ]
  },
  "code": 403
}]
F0224 13:11:02.660565    4303 helpers.go:114] Error from server (Forbidden): unknown (get pods)
```

The proxy server's logs on the `example-east` (leaf) cluster are empty (which indicates that the request probably never gets sent over)

Here's an excerpt from the proxy server's log on the `example-main` (root) cluster:

```bash
INFO [RBAC]      Access to create user in namespace default denied to roles Proxy,default-implicit-role: no allow rule matched. services/role.go:1891
WARN [PROXY:KUB] this user cannot request kubernetes access, has no assigned groups proxy/forwarder.go:322
INFO [RBAC]      Access to create user in namespace default denied to roles Proxy,default-implicit-role: no allow rule matched. services/role.go:1891
WARN [PROXY:KUB] this user cannot request kubernetes access, has no assigned groups proxy/forwarder.go:322
INFO [RBAC]      Access to create user in namespace default denied to roles Proxy,default-implicit-role: no allow rule matched. services/role.go:1891
WARN [PROXY:KUB] this user cannot request kubernetes access, has no assigned groups proxy/forwarder.go:322
INFO [RBAC]      Access to create user in namespace default denied to roles Proxy,default-implicit-role: no allow rule matched. services/role.go:1891
WARN [PROXY:KUB] this user cannot request kubernetes access, has no assigned groups proxy/forwarder.go:322
INFO [RBAC]      Access to create user in namespace default denied to roles Proxy,default-implicit-role: no allow rule matched. services/role.go:1891
WARN [PROXY:KUB] this user cannot request kubernetes access, has no assigned groups proxy/forwarder.go:322
INFO [RBAC]      Access to create user in namespace default denied to roles Proxy,default-implicit-role: no allow rule matched. services/role.go:1891
WARN [PROXY:KUB] this user cannot request kubernetes access, has no assigned groups proxy/forwarder.go:322
```

It appears that the lack of any entry other than `{{internal.kubernetes_groups}}` on the **root** cluster's role definition prevents the group from being applied on the **leaf** end.

This can be worked around by changing the `dev` role on the **root** cluster and adding a 'dummy' group like so:

```yaml
    kubernetes_groups:
    - '{{internal.kubernetes_groups}}'
    - dummy-group
```

Once this role is applied, Kubernetes access to the leaf cluster works fine:

`example-main` (root) logs:
```
DEBU [PROXY:KUB] Client certificate of gus@gravitational.com has requested routing to a cluster: example-east.gravitational.co. proxy/forwarder.go:380
DEBU [PROXY:KUB] Requesting new creds for user: gus@gravitational.com, groups: [dummy-group system:authenticated], cluster: example-east.gravitational.co. proxy/forwarder.go:822
DEBU [PROXY:KUB] Requesting K8s cert for user: gus@gravitational.com, groups: [dummy-group system:authenticated], cluster: example-east.gravitational.co. proxy/forwarder.go:983
DEBU [AUTH]      Generating certificate to access remote Kubernetes clusters. auth/kube.go:81
INFO [CA]        Generating TLS certificate {0x373d1a0 0xc00160ee00 CN=gus@gravitational.com,OU=usage:kube,O=dev,POSTALCODE=null,STREET=,ST=dummy-group 2020-02-25 05:10:38.377874203 +0000 UTC []}. common_name:gus@gravitational.com dns_names:[] locality:[] not_after:2020-02-25 05:10:38.377874203 +0000 UTC org:[dev] org_unit:[usage:kube] tlsca/ca.go:203
DEBU [PROXY:KUB] Received valid K8s cert for user: gus@gravitational.com, groups: [dummy-group system:authenticated], cluster: example-east.gravitational.co. proxy/forwarder.go:1018
DEBU [PROXY:KUB] Created new session for user: gus@gravitational.com, groups: [dummy-group system:authenticated], cluster: example-east.gravitational.co. proxy/forwarder.go:930
DEBU [PROXY:SER] Dialing from 1.2.3.4:55298 to remote.kube.proxy.teleport.cluster.local. example-east.gravitational.co reversetunnel/remotesite.go:521
DEBU [PROXY:SER] Requesting connection to remote.kube.proxy.teleport.cluster.local [] in remote cluster. example-east.gravitational.co reversetunnel/remotesite.go:585
```

`example-east` (leaf) logs:
```
DEBU [AUTH:1]    Server certificate cert(cb957659-cbb2-471f-8ad0-5d3984be1563.example-east.gravitational.co issued by example-east.gravitational.co:271721871327075382353220156241970883518). auth/middleware.go:164
DEBU [AUTH]      Mapped roles [dev] of remote user "gus@gravitational.com" to local roles [admin] and traits map[kubernetes_groups:[dummy-group] logins:[]]. auth/permissions.go:159
DEBU [PROXY:KUB] Impersonate User: User(name=remote-gus@gravitational.com-example-main.gravitational.co, roles=[admin], identities=[]) proxy/forwarder.go:658
DEBU [PROXY:KUB] Impersonate Group: dummy-group proxy/forwarder.go:658
DEBU [PROXY:KUB] Impersonate Group: system:masters proxy/forwarder.go:658
DEBU [PROXY:KUB] Impersonate Group: system:authenticated proxy/forwarder.go:658
DEBU [PROXY:KUB] Using Bearer Token Auth proxy/forwarder.go:664
DEBU [PROXY:SER] Dialing from 127.0.0.1:50958 to 10.7.240.1:443. trace.fields:map[cluster:example-east.gravitational.co] reversetunnel/localsite.go:189
DEBU [HTTP:PROX] No valid environment variables found. proxy/proxy.go:217
DEBU [HTTP:PROX] No proxy set in environment, returning direct dialer. proxy/proxy.go:137
INFO             Round trip: GET https://10.7.240.1:443/api/v1/namespaces/default/pods?limit=500, code: 200, duration: 19.509186ms tls:version: 304, tls:resume:false, tls:csuite:1301, tls:server:remote.kube.proxy.teleport.cluster.local forward/fwd.go:196
```

Client output:
```
$ kubectl config get-contexts
CURRENT   NAME                            CLUSTER                                                   AUTHINFO                                              NAMESPACE
*         example-east.gravitational.co   example-east.gravitational.co                             example-east.gravitational.co                             
          example-main.gravitational.co   example-main.gravitational.co                             example-main.gravitational.co          

$ kubectl get pods
NAME                        READY   STATUS      RESTARTS   AGE
kaniko                      0/1     Completed   0          283d
registry-7cb69cbf59-trbn9   1/1     Running     0          9d
```

If this functionality worked correctly in the past, then this is a regression.

There is a similar bug reported by another customer which affects `{{internal.logins}}` being mapped over to trusted clusters in a similar way. I'll report this separately.

**What you expected to happen**: The kubernetes connection to the remote cluster should be made as the `system:masters` group, as this is contained in the leaf cluster's role defintion.

**How to reproduce it (as minimally and precisely as possible)**: Cluster setup details (the only relevant parts really are the uses of `{{internal.kubernetes_groups}}`)

**root** cluster, `dev` role:
```yaml
kind: role
version: v3
metadata:
  name: dev
spec:
  options:
    cert_format: standard
    forward_agent: true
    max_session_ttl: 30h0m0s
    port_forwarding: true
  allow:
    kubernetes_groups:
    - '{{internal.kubernetes_groups}}'
    logins:
    - '{{internal.logins}}'
    - root
    node_labels:
      'role': 'node'
    rules:
    - resources:
      - session
      verbs:
      - list
      - read
  deny:
    logins: null
    rules:
    - resources:
      - role
      verbs:
      - list
      - create
      - read
      - update
      - delete
    - resources:
      - auth_connector
      verbs:
      - list
      - create
      - read
      - update
      - delete
    - resources:
      - trusted_cluster
      verbs:
      - list
      - create
      - read
      - update
      - delete
    - resources:
      - event
      verbs:
      - list
      - read
```

**leaf** cluster, `admin` role:
```yaml
kind: role
metadata:
  id: 1580804462883211642
  name: admin
spec:
  allow:
    kubernetes_groups:
    - '{{internal.kubernetes_groups}}'
    - system:masters
    logins:
    - '{{internal.logins}}'
    - root
    node_labels:
      '*': '*'
    rules:
    - resources:
      - role
      verbs:
      - list
      - create
      - read
      - update
      - delete
    - resources:
      - auth_connector
      verbs:
      - list
      - create
      - read
      - update
      - delete
    - resources:
      - session
      verbs:
      - list
      - read
    - resources:
      - trusted_cluster
      verbs:
      - list
      - create
      - read
      - update
      - delete
    - resources:
      - event
      verbs:
      - list
      - read
  deny:
    logins: null
  options:
    cert_format: standard
    forward_agent: true
    max_session_ttl: 30h0m0s
    port_forwarding: true
version: v3
```

**leaf** cluster, trusted cluster definition:

```yaml
kind: trusted_cluster
metadata:
  id: 1582559481366881536
  name: example-main.gravitational.co
spec:
  enabled: true
  role_map:
    - remote: "*"
      local: ["admin"] 
  token: <redacted>
  tunnel_addr: example-main.gravitational.co:3024
  web_proxy_addr: example-main.gravitational.co:3080
version: v2
```

**Environment**:
- Teleport version (use `teleport version`): `Teleport Enterprise v4.2.2git:v4.2.2-0-gb06a05d2 go1.13.2`
- Tsh version (use `tsh version`): `Teleport v4.2.2 git:v4.2.2-0-gb06a05d2 go1.13.2`
- OS (e.g. from /etc/os-release): Fedora 30

**Comment by thejosephstevens** [comment]
Did this ever get fixed? I'm running into a similar issue on Teleport 4.3.0 with GKE. Not sure how deep I should be going on debugging.

**Comment by webvictim** [comment]
No, this hasn't been fixed.

**Comment by awly** [comment]
This will be fixed in teleport 5.2

**Comment by awly** [comment]
I'm going to believe my past self and mark this as fixed, since we're way past 5.2 now

**Comment by Rishav-Zn** [comment]
I am trying to do the same as mentioned in the issue, i have added the leaf cluster successfully . I have access from root to leaf for my admins account but unable to map role based access for particular user for a certain ssh resource.
